src/instances.ts

```diff
@@ -266,17 +272,25 @@ function successfulTypeScriptInstance(
       .getProgram()
       .getProgram();
   } else {
-    const servicesHost = makeServicesHost(scriptRegex, log, loader, instance);
+    const cachedServicesHost = makeServicesHost(
+      scriptRegex,
+      log,
+      loader,
+      instance,
+      loaderOptions.experimentalFileCaching
+    );
+
     instance.languageService = compiler.createLanguageService(
-      servicesHost,
+      cachedServicesHost.servicesHost,
       compiler.createDocumentRegistry()
     );
-  }
 
-  if (!loader._compiler.hooks) {
-    throw new Error(
-      "You may be using an old version of webpack; please check you're using at least version 4"
-    );
+    if (cachedServicesHost.clearCache !== null) {
+      loader._compiler.hooks.watchRun.tap(
```

Comment:
I just wanted to check; is `watchRun` definitely `sync` not `async`?

Reply:
It seems that `watchRun` is Tappable, which supports both `async` and `sync` methods (`tap` and `tapAsync`).

https://webpack.js.org/api/compiler-hooks/#watchrun
https://github.com/webpack/tapable/blob/04e52118691e0a6c77a42d15e99dc4bd2bfca07c/lib/AsyncSeriesHook.js#L21

Reply:
Is there any advantage to using `tapAsync` instead of `tap`?  I've an idea it doesn't really matter but I thought I'd check.

Reply:
I believe that it depends on what are you doing in your handler. In this case we need to use sync because we need to just clear cache without any async operation. But if you find something interesting about it - please let me know 🙂